src/chainstate/coordinator/mod.rs

```diff
@@ -623,7 +1263,42 @@ impl<
 
             // at this point, we need to figure out if the sortition we are
             //  about to process is the first block in reward cycle.
-            let reward_cycle_info = self.get_reward_cycle_info(&header)?;
+            let mut reward_cycle_info = self.get_reward_cycle_info(&header)?;
+
+            if let Some(rc_info) = reward_cycle_info.as_mut() {
+                let cur_epoch =
+                    SortitionDB::get_stacks_epoch(self.sortition_db.conn(), header.block_height)?
+                        .expect(&format!(
+                            "BUG: no epoch defined at height {}",
+                            header.block_height
+                        ));
+
+                if cur_epoch.epoch_id >= StacksEpochId::Epoch21 {
+                    // potentially have an anchor block, but only process the next reward cycle (and
+                    // subsequent reward cycles) with it if the prepare-phase block-commits affirm its
+                    // presence.  This only gets checked in Stacks 2.1 or later.
+
+                    // NOTE: this mutates rc_info
+                    if let Some(missing_anchor_block) = self
+                        .reinterpret_affirmed_pox_anchor_block_status(
+                            &canonical_affirmation_map,
+                            &header,
+                            rc_info,
+                        )?
+                    {
+                        // missing this anchor block -- cannot proceed
+                        info!("Burnchain block processing stops due to missing affirmed anchor block {}", &missing_anchor_block);
+                        return Ok(Some(missing_anchor_block));
+                    }
+                }
```

Comment:
I think these checks should be applied retroactively. In the event of an 1805 related reorg, it would almost certainly rise to the need for a SIP-011 procedure, so making this apply to earlier reward cycles will almost surely lead to no change in consensus. If it would, then `next` could always be modified to undo the retroactive application of affirmation maps.

I think it's very important to testing that this code be applied retroactively -- it would give the network the assurance that this code can correctly handle PoX as it exists today when all is going well.

Reply:
I've made this a config option in the node.